Fix implementation of cumsum and diff function on judiVector to be dimensional

ziyiyin97 opened the pull request:
I gave a second thought to the implementation of cumsum and diff on judiVector and I thought they should be dimensional, just like what norm and inner product do. So, depending on integration along time or along receivers, I multiply the term `dt` or `drec` to the cumsum result. Similarly, in the case of differentiation, I divide by the term `dt` or `drec`.

Also, there was a problem on diff function while dims=2. I fix that in this PR and add more associated tests to thoroughly test every case.

Appreciate any advice and review!

mloubout commented on src/TimeModeling/judiVector.jl:
```diff
@@ -713,19 +713,25 @@ iterate(S::judiVector, state::Integer=1) = state > S.nsrc ? nothing : (S.data[st
 function cumsum(x::judiVector;dims=1)
     dims == 1 || dims == 2 || throw(judiVectorException("Dimension $(dims) is out of range for a 2D array"))
-    y = deepcopy(x)
+    y = (dims == 1 ? x.geometry.dt[1] : (x.geometry.xloc[1][2]-x.geometry.xloc[1][1])) * x
```
In addition to what felix said on slack, this would only work for uniformly sampled receiver along the x coordinate only in 2D which is very restrictive.

mloubout commented on src/TimeModeling/judiVector.jl:
```diff
@@ -713,19 +713,25 @@ iterate(S::judiVector, state::Integer=1) = state > S.nsrc ? nothing : (S.data[st
-    y = deepcopy(x)
-    for i = 1:x.nsrc
-        y.data[i][2:end,:] = diff(x.data[i],dims=dims)
+    y = 1/(dims == 1 ? x.geometry.dt[1] : (x.geometry.xloc[1][2]-x.geometry.xloc[1][1])) * x
```
same

mloubout commented on src/TimeModeling/judiVector.jl:
```diff
@@ -713,19 +713,25 @@ iterate(S::judiVector, state::Integer=1) = state > S.nsrc ? nothing : (S.data[st
-    for i = 1:x.nsrc
-        y.data[i][2:end,:] = diff(x.data[i],dims=dims)
+    y = 1/(dims == 1 ? x.geometry.dt[1] : (x.geometry.xloc[1][2]-x.geometry.xloc[1][1])) * x
+    if dims == 1
```
Unnecesseary if/else

mloubout replied:
ping

ziyiyin97 replied:
This if/else is needed because diff works differently for dims=1 and dims=2. We should keep the first column/row of the matrix based on different cases.

mloubout replied:
use `selectdim` for that

ziyiyin97 replied:
Done

ziyiyin97 commented:
Thanks for the feedback! Based on our discussion, I only leave the time derivative/integration and report exception on other dimensions because receivers might not be uniform spaced. Also fix the tests.

mloubout commented on src/TimeModeling/judiVector.jl:
```diff
@@ -712,20 +712,20 @@ iterate(S::judiVector, state::Integer=1) = state > S.nsrc ? nothing : (S.data[st
 function cumsum(x::judiVector;dims=1)
-    dims == 1 || dims == 2 || throw(judiVectorException("Dimension $(dims) is out of range for a 2D array"))
-    y = deepcopy(x)
+    dims == 1 || throw(judiVectorException("Only time integration is supported"))
```
No, the second dim should still be supported just adimensional, same below.

mloubout replied:
And add test for dim=2

mloubout commented on test/test_judiVector.jl:
```diff
@@ -359,16 +359,27 @@ ftol = 1e-6
+    d_orig = judiVector(rec_geometry, refarray)
+
+    dt = rec_geometry.dt[1]
+    drec = rec_geometry.xloc[1][2]-rec_geometry.xloc[1][1]
```
unused

ziyiyin97 commented:
Thanks for your feedback. Fix cumsum/diff on 2nd (receiver) to be non-dimensional, e.g. sample interval = 1.

mloubout commented on src/TimeModeling/judiVector.jl:
```diff
@@ -713,19 +713,25 @@ iterate(S::judiVector, state::Integer=1) = state > S.nsrc ? nothing : (S.data[st
+    y = (dims == 1 ? x.geometry.dt[1] : 1f0) * x        # receiver dimension is non-dimensional
     for i = 1:x.nsrc
-        cumsum!(y.data[i], x.data[i], dims=dims)
+        cumsum!(y.data[i], y.data[i], dims=dims)
```
Yu can't do that, that's not in place and will create a temporary array. Most in-place operation do not allow to alias the result to the input (you should check it up), so you can have the result be the same object as the input. so like 716 should only get ` h = dims == 1 ? x.geometry.dt[1] : 1f0` then `y = similar(x)` then `cumsum!(y, x); y .*= h` (`rmul!` even better bit not implemented I think)

ziyiyin97 commented:
Thanks for your suggestions! To make it clean,

1. I overloaded `lmul!, rmul!, ldiv!, rdiv!` functions for judiVector so that they could be easily deployed for `cumsum`.
2. I modify scalar `*` and `/` operations so that they don't have either to loop over shots or to allocate extra memory.
3. I modified cumsum function as such.
4. I added associated tests with them in `test_judiVector.jl`.

mloubout commented on src/TimeModeling/judiVector.jl:
```diff
@@ -351,29 +351,60 @@ end
+function ldiv!(b::Number, a::judiVector{avDT}) where avDT
+    typeof(a.data[1]) == SeisCon && throw("Division for OOC judiVectors not supported.")
+    if iszero(b)
+        error("Division by zero")
```
@throw ....

mloubout replied:
Actually, always attach an Error type to a `throw` same for OOC throws

mloubout replied:
https://docs.julialang.org/en/v1/manual/control-flow/

There is the list of built-ins exceptions there

ziyiyin97 commented:
Great suggestions! I've modified the error messages for +-*/ operations.
Division by zero is classified as a `DivideError()`
Operations with OOC object is classified as a `DomainError()`.

ziyiyin97 commented:
`ldiv!, rdiv!` are not supported in Julia 1.1 version. Thus, switch to a `try/catch` framework.

mloubout commented on src/TimeModeling/judiVector.jl:
```diff
@@ -351,30 +351,62 @@ end
+    catch e             # julia 1.1 does not support ldiv!
+        lmul!(1f0/b, a.data)
+        return a
+    end
```
That's quite an overkill, just do`lmul!`

ziyiyin97 commented:
Modified `ldiv, rdiv` functions for judiVector. Also remove `test_wavefield.jl` file from `src/TimeModeling/` folder since I don't think it is used anywhere.

ziyiyin97 commented:
Use `selectdim` in `diff` which makes it cleaner with dimensions and avoids extra space allocation

ziyiyin97 commented on test/test_judiVector.jl:
```diff
@@ -134,6 +134,20 @@ ftol = 1e-6
+    @test iszero(norm(2f0*d_obs - d1))
+    @test iszero(norm(3f0*d_obs - d2))
+    @test iszero(norm(d_obs/4f0 - d3))
+    @test iszero(norm(d_obs/5f0 - d4))
```
Of course!

ziyiyin97 commented:
Modify tests for `lmul, rmul, ldiv, rdiv` to be more explicit.
